src/main/java/gregtech/common/covers/CoverPump.java

```diff
@@ -105,37 +116,68 @@ protected int doTransferFluids(int transferLimit) {
-        } else if (pumpMode == PumpMode.EXPORT) {
-            return moveHandlerFluids(myFluidHandler, fluidHandler, transferLimit, fluidFilter::testFluidStack);
+    protected int doTransferFluidsInternalSmart(IFluidHandler coveredBlockFluidHandler, IFluidHandler connectionFluidHandler, int transferLimit) {
+        switch (this.pumpMode) {
+            case EXPORT: return doExportFluid(coveredBlockFluidHandler , connectionFluidHandler, transferLimit);
+            case IMPORT: return doImportFluid(coveredBlockFluidHandler , connectionFluidHandler, transferLimit);
+            default:
+                return 0;
         }
-        return 0;
     }
 
-    public static int moveHandlerFluids(IFluidHandler sourceHandler, IFluidHandler destHandler, int transferLimit, Predicate<FluidStack> fluidFilter) {
+    protected int doImportFluid(IFluidHandler me, IFluidHandler dest, int transferLimit) {
         int fluidLeftToTransfer = transferLimit;
-        for (IFluidTankProperties tankProperties : sourceHandler.getTankProperties()) {
-            FluidStack currentFluid = tankProperties.getContents();
-            if (currentFluid == null || currentFluid.amount == 0 || !fluidFilter.test(currentFluid)) continue;
-            currentFluid.amount = fluidLeftToTransfer;
-            FluidStack fluidStack = sourceHandler.drain(currentFluid, false);
-            if (fluidStack == null || fluidStack.amount == 0) continue;
-            int canInsertAmount = destHandler.fill(fluidStack, false);
-            if (canInsertAmount > 0) {
-                fluidStack.amount = canInsertAmount;
-                fluidStack = sourceHandler.drain(fluidStack, true);
-                if (fluidStack != null && fluidStack.amount > 0) {
-                    destHandler.fill(fluidStack, true);
-
-                    fluidLeftToTransfer -= fluidStack.amount;
-                    if (fluidLeftToTransfer == 0) break;
-                }
+        for (IFluidTankProperties sourceTankProperties : dest.getTankProperties()) {
+
+            FluidStack currentFluid = sourceTankProperties.getContents();
+            if (currentFluid == null || currentFluid.amount == 0 || !fluidFilter.testFluidStack(currentFluid)) continue;
+
+            FluidStack currentFluidCopy = currentFluid.copy();
+            currentFluidCopy.amount = fluidLeftToTransfer;
+            FluidStack canExtractFluid = dest.drain(currentFluidCopy, false);
+            if (canExtractFluid == null || canExtractFluid.amount == 0) continue;
+            int canInsertAmount = me.fill(canExtractFluid, false);
+            if (canInsertAmount == 0) continue;
+            int liquidDeltaLimitFromPortion = Math.round((sourceTankProperties.getCapacity() * ((float) this.percentageInTank / 100))) - sourceTankProperties.getContents().amount;
+            int finalTransferAmount = Math.min(liquidDeltaLimitFromPortion, canInsertAmount);
```

Comment:
Maximal liquid amount that can be transferred should be calculated from `transferLimit` before even attempting draining fluid as dry run because there may be no need for it as tank already has enough fluid.

Reply:
Added `if (fluidLeftToTransfer == 0) break;` before drain